include/seastar/util/memory-data-source.hh

```diff
@@ -0,0 +1,67 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2025 ScyllaDB
+ */
+
+#pragma once
+
+#include <seastar/core/iostream.hh>
+
+namespace seastar {
+namespace util {
+
+template <typename Container>
+requires std::is_nothrow_move_constructible_v<Container>
+class memory_data_source final : public data_source_impl {
```

Comment:
Suggest to call this basic_memory_data_source and define basic_memory_data_source\<std::vector\<temporary_buffer\<char>>> as memory_data_source.

Need more constraints.

Do we want to export it? I guess it's generally useful.

Option: make the container parameter a template template parameter, so you can write memory_data_source\<std::vector>.